packages/grid/x-data-grid/src/hooks/features/density/useGridDensity.tsx

```diff
@@ -18,51 +21,114 @@ const getUpdatedDensityState = (
   newDensity: GridDensity,
   newHeaderHeight: number,
   newRowHeight: number,
+  newMaxDepth: number,
+  newHeaderGroupingRowHeight: number,
 ): GridDensityState => {
   switch (newDensity) {
     case GridDensityTypes.Compact:
       return {
         value: newDensity,
         headerHeight: Math.floor(newHeaderHeight * COMPACT_DENSITY_FACTOR),
         rowHeight: Math.floor(newRowHeight * COMPACT_DENSITY_FACTOR),
+        headerGroupingRowHeight: Math.floor(newHeaderGroupingRowHeight * COMPACT_DENSITY_FACTOR),
+        headerGroupingMaxDepth: newMaxDepth,
         factor: COMPACT_DENSITY_FACTOR,
       };
     case GridDensityTypes.Comfortable:
       return {
         value: newDensity,
         headerHeight: Math.floor(newHeaderHeight * COMFORTABLE_DENSITY_FACTOR),
         rowHeight: Math.floor(newRowHeight * COMFORTABLE_DENSITY_FACTOR),
+        headerGroupingRowHeight: Math.floor(
+          newHeaderGroupingRowHeight * COMFORTABLE_DENSITY_FACTOR,
+        ),
+        headerGroupingMaxDepth: newMaxDepth,
         factor: COMFORTABLE_DENSITY_FACTOR,
       };
     default:
       return {
         value: newDensity,
         headerHeight: newHeaderHeight,
         rowHeight: newRowHeight,
+        headerGroupingRowHeight: newHeaderGroupingRowHeight,
+        headerGroupingMaxDepth: newMaxDepth,
         factor: 1,
       };
   }
 };
 
 export const densityStateInitializer: GridStateInitializer<
-  Pick<DataGridProcessedProps, 'density' | 'headerHeight' | 'rowHeight'>
-> = (state, props) => ({
-  ...state,
-  density: getUpdatedDensityState(props.density, props.headerHeight, props.rowHeight),
-});
+  Pick<
+    DataGridProcessedProps,
+    'density' | 'headerHeight' | 'rowHeight' | 'headerGroupingRowHeight' | 'columnGroupingModel'
+  >
+> = (state, props) => {
+  // TODO: think about improving this initialization. Could it be done in the useColumn initializer?
+  // TODO: manage to remove ts-ignore
+  let maxDepth: number;
+  if (props.columnGroupingModel == null || Object.keys(props.columnGroupingModel).length === 0) {
+    maxDepth = 0;
+  } else {
+    const unwrappedGroupingColumnModel = unwrapGroupingColumnModel(props.columnGroupingModel);
+
+    const visibleColumns = state.columns!.all!.filter(
+      (field) => state.columns!.columnVisibilityModel![field!] !== false,
+    );
+
+    if (visibleColumns.length === 0) {
+      maxDepth = 0;
+    } else {
+      maxDepth = Math.max(
+        ...visibleColumns.map((field) => unwrappedGroupingColumnModel[field!]?.length ?? 0),
```

Comment:
This looks a bit fragile. The state is partial at this point, so TS has no guarantee it's already there.
We kind of know that columns state is there because of the order of the hooks.

More pretty way to do the same avoiding `!` on each step below, but I'm not sure it's the way to go:

```suggestion
    const columnsState = state.columns as GridStateCommunity['columns'];
    const visibleColumns = columnsState.all.filter(
      (field) => columnsState.columnVisibilityModel[field] !== false,
    );

    if (visibleColumns.length === 0) {
      maxDepth = 0;
    } else {
      maxDepth = Math.max(
        ...visibleColumns.map((field) => unwrappedGroupingColumnModel[field]?.length ?? 0),
```

cc @flaviendelangle

Reply:
I also prefer this single cast
As for the fragile aspect, in the future it would be nice to be able to type a partial state, something like `GridState<PluginAlreadyLoaded1, PluginAlreadyLoaded2`.
But that's potentially a lot of work.